File: apps/ios/ChatApp/Sources/Services/APIClient.swift
import Foundation

enum APIError: Error, LocalizedError {
    case http(Int, String)
    case transport(Error)

    var errorDescription: String? {
        switch self {
        case .http(let code, let body): return "HTTP \(code): \(body)"
        case .transport(let err): return err.localizedDescription
        }
    }
}

struct APIClient {
    static let baseURL: String = {
        Bundle.main.object(forInfoDictionaryKey: "CHATAPP_API_BASE") as? String
            ?? ProcessInfo.processInfo.environment["CHATAPP_API_BASE"]
            ?? "http://localhost:8080"
    }()

    static let wsBaseURL: String = {
        Bundle.main.object(forInfoDictionaryKey: "CHATAPP_WS_BASE") as? String
            ?? ProcessInfo.processInfo.environment["CHATAPP_WS_BASE"]
            ?? "ws://localhost:8080"
    }()

    static let mediaBaseURL: String = {
        Bundle.main.object(forInfoDictionaryKey: "CHATAPP_MEDIA_BASE") as? String
            ?? ProcessInfo.processInfo.environment["CHATAPP_MEDIA_BASE"]
            ?? "http://localhost:8100"
    }()

    let token: String?

    init(token: String? = nil) {
        self.token = token
    }

    // Signed-grant media upload matching the web/Android flow: fetch a
    // short-lived upload token from the Go API, then POST the raw bytes to
    // the C++ media edge. Returns the absolute media URL.
    func uploadMedia(filename: String, data: Data) async throws -> String {
        var grant = ""
        if let t = try? await post("/api/media/upload-token"),
           let obj = try JSONSerialization.jsonObject(with: t) as? [String: Any],
           let exp = obj["expires"] as? Int, let sig = obj["signature"] as? String {
            let allowed = CharacterSet.urlQueryAllowed
            grant = "&exp=\(exp)&sig=\(sig.addingPercentEncoding(withAllowedCharacters: allowed) ?? sig)"
        }
        let name = filename.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? filename
        var req = URLRequest(url: URL(string: "\(Self.mediaBaseURL)/upload?filename=\(name)\(grant)")!)
        req.httpMethod = "POST"
        req.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
        req.httpBody = data
        let (respData, resp) = try await URLSession.shared.data(for: req)
        guard (200..<300).contains((resp as? HTTPURLResponse)?.statusCode ?? -1),
              let obj = try JSONSerialization.jsonObject(with: respData) as? [String: Any],
              let rel = obj["url"] as? String else {
            throw APIError.http((resp as? HTTPURLResponse)?.statusCode ?? -1, "upload failed")
        }
        return Self.mediaBaseURL + rel
    }

    func get(_ path: String) async throws -> Data {
        try await request(path, method: "GET")
    }

    func post(_ path: String, body: [String: Any] = [:]) async throws -> Data {
        try await request(path, method: "POST", body: body)
    }

    func put(_ path: String, body: [String: Any] = [:]) async throws -> Data {
        try await request(path, method: "PUT", body: body)
    }

    func delete(_ path: String) async throws -> Data {
        try await request(path, method: "DELETE")
    }

    func delete(_ path: String, body: [String: Any]) async throws -> Data {
        try await request(path, method: "DELETE", body: body)
    }

    private func request(_ path: String, method: String, body: [String: Any]? = nil) async throws -> Data {
        var req = URLRequest(url: URL(string: Self.baseURL + path)!)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token { req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        if let body { req.httpBody = try JSONSerialization.data(withJSONObject: body) }
        do {
            let (data, resp) = try await URLSession.shared.data(for: req)
            let status = (resp as? HTTPURLResponse)?.statusCode ?? -1
            guard (200..<300).contains(status) else {
                throw APIError.http(status, String(data: data, encoding: .utf8) ?? "")
            }
            return data
        } catch let error as APIError {
            throw error
        } catch {
            throw APIError.transport(error)
        }
    }
}

File: apps/android/app/src/main/java/com/chatapp/data/ApiClient.kt
package com.chatapp.data

import java.io.IOException
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

// Thin OkHttp wrapper matching the Go API contract: JSON bodies, bearer
// tokens, string error payload from {"error": ...}. Callers run these on
// Dispatchers.IO.
class ApiClient(private val baseUrl: String) {

    private val client = OkHttpClient()
    private val jsonMedia = "application/json".toMediaType()

    fun get(path: String, token: String? = null): String =
        execute(newRequest(path, token).get().build())

    fun post(path: String, body: String = "{}", token: String? = null): String =
        execute(newRequest(path, token).post(body.toRequestBody(jsonMedia)).build())

    fun put(path: String, body: String = "{}", token: String? = null): String =
        execute(newRequest(path, token).put(body.toRequestBody(jsonMedia)).build())

    fun delete(path: String, body: String = "{}", token: String? = null): String =
        execute(newRequest(path, token).delete(body.toRequestBody(jsonMedia)).build())

    // Signed-grant media upload matching the web flow: fetch a short-lived
    // upload token from the Go API, then POST the raw bytes to the C++ media
    // edge. Returns the absolute media URL.
    fun uploadMedia(mediaBase: String, filename: String, bytes: ByteArray, token: String): String {
        var grant = ""
        try {
            val t = org.json.JSONObject(post("/api/media/upload-token", "{}", token))
            grant = "&exp=${t.getLong("expires")}&sig=${java.net.URLEncoder.encode(t.getString("signature"), "UTF-8")}"
        } catch (_: Exception) { /* dev mode: unsigned upload accepted */ }
        val url = "$mediaBase/upload?filename=${java.net.URLEncoder.encode(filename, "UTF-8")}$grant"
        val req = Request.Builder().url(url)
            .post(bytes.toRequestBody("application/octet-stream".toMediaType())).build()
        val resp = execute(req)
        val rel = org.json.JSONObject(resp).getString("url")
        return "$mediaBase$rel"
    }

    private fun newRequest(path: String, token: String?): Request.Builder {
        val builder = Request.Builder()
            .url("$baseUrl$path")
            .header("Content-Type", "application/json")
        if (!token.isNullOrEmpty()) builder.header("Authorization", "Bearer $token")
        return builder
    }

    private fun execute(request: Request): String {
        client.newCall(request).execute().use { resp ->
            val body = resp.body?.string().orEmpty()
            if (!resp.isSuccessful) {
                val msg = try {
                    org.json.JSONObject(body).optString("error").ifEmpty { "HTTP ${resp.code}" }
                } catch (_: Exception) { "HTTP ${resp.code}" }
                throw IOException(msg)
            }
            return body
        }
    }
}
